api/src/main/java/com/graphhopper/routing/util/CustomModel.java

```diff
@@ -0,0 +1,269 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.graphhopper.routing.util;
+
+import com.graphhopper.json.geo.JsonFeature;
+import com.graphhopper.util.Parameters;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class is used in combination with CustomProfileConfig.
+ */
+public class CustomModel {
+
+    public static final String KEY = "custom_model";
+
+    static double DEFAULT_D_I = 70;
+    // optional:
+    private Double maxSpeedFallback, vehicleWeight, vehicleWidth, vehicleHeight, vehicleLength;
+    private Double headingPenalty = Parameters.Routing.DEFAULT_HEADING_PENALTY;
+    // default value derived from the cost for time e.g. 25€/hour and for distance 0.5€/km, for trucks this is usually larger
+    private double distanceInfluence = DEFAULT_D_I;
+    private Map<String, Object> speedFactor = new HashMap<>();
+    private Map<String, Object> maxSpeed = new HashMap<>();
+    private Map<String, Object> priorityMap = new HashMap<>();
+    private Map<String, JsonFeature> areas = new HashMap<>();
+
+    public CustomModel() {
+    }
+
+    public CustomModel(CustomModel toCopy) {
+        this.maxSpeedFallback = toCopy.maxSpeedFallback;
+        this.vehicleWeight = toCopy.vehicleWeight;
+        this.vehicleWidth = toCopy.vehicleWidth;
+        this.vehicleLength = toCopy.vehicleLength;
+        this.vehicleHeight = toCopy.vehicleHeight;
```

Comment:
I wonder if we should group these into a subnode (like vehicleDimensions), because it seems a bit distracting that they are all at the root level. speed_factor, max_speed and priority seem more important.

Reply:
good idea

Reply:
Or should this even be more general, like:

```yaml
priority:
  thresholds: {max_weight: 3.5, max_length: 2.1}
```

Simply typing into your custom model how large you vehicle is is very handy, but I wonder how long it is going to take until we have other numeric encoded values that we'd like to use in a similar way. Actually there is alread 'max_axle_load'...

Reply:
Well max_axle_load should fit into vehicle dimensions... Do you think the numeric encoded values will always be like max_something and then we can just specify what something of our vehicle is?

Reply:
> Do you think the numeric encoded values will always be like max_something 

There are other use cases like a value that reflects the priority directly 

> and then we can just specify what something of our vehicle is?

What do you mean here?

Reply:
> Simply typing into your custom model how large you vehicle is is very handy

The problem with putting these values inside the priority (or other maps) is that 
1. the numeric value is now used for the limit (excluding in `priority` is normally done via value==0) and that
2. you do not know if this is the street max width or the vehicle max width

Reply:
> There are other use cases like a value that reflects the priority directly

Ok but how are they supposed to fit in then?

> What do you mean here?
Sorry, I meant so far we use encoded values like `max_weight` and `max_length` which all specify the maximum value some vehicle property might have so the vehicle can still drive on a route. That is why in the custom model we simply define the properties of the **vehicle** and this is then used to decide whether or not this vehicle can use a certain road. But this is the only way of using numeric encoded values for custom weighting so far and I was wondering if this will always be so (but apparently you already can think of other cases).

Reply:
> Sorry, I meant so far we use encoded values like max_weight and max_length

Ah, no. It is good that we discuss and "brainstorm" about this. So how to go forward after this is a really interesting question :)

> Ok but how are they supposed to fit in then?

Don't know for sure, but one way could be to use them directly where we would use the number, so instead of:

```yaml
priority: { road_class: { motorway: 0.5 } }
```

it could be

```yaml
priority: { road_class: { motorway: my_ev } }
```

or if we would want to apply them unconditionally it could be just:

```yaml
priority:
   road_class: { motorway: 0.5 }
   my_ev2
```

Or if this is not valid yaml we could do it more explicitly:

```yaml
priority:
   road_class: { motorway: 0.5 }
   "*": my_ev2
```

Reply:
Ah ok some great suggestions her :+1: But you mean this is for 2.0 then? For this release I still wonder if we can get rid of the hard-coded max_weight/vehicle/length rules (see slack)

Reply:
Ah, yes. Let's keep here probably with discussion ;)
We could put "thresholds" into the root element or create 2 new root elements block_if_bigger and block_if_smaller:

```yaml
block_if_bigger:
   max_weight: 2.5
```

Reply:
Another generic solution for integer and decimal values could be:

```
priority:
   max_weight: { ">2.5": 0 }
```

Reply:
So as soon as we stumble over an IntegerEncodedValue instead of an EnumEncodedValue we could support `<` and `>`? (and without any operand it is `==` like for enums)

Reply:
> Another generic solution for integer and decimal values could be

I really like this, because it follows the same pattern as the enum rules and therefore makes it easy to understand. It could also be extended even more to multiple ranges like `max_weight: { "[0,10]": 0.3, "[30,50]": 0.7}`.

Btw if we want to apply the same logic also for boolean, it would have to be `get_off_bike: {"true": 0.4, "false": 0.2}`.

Reply:
So if you ask me this is how we should do it, the only thing I am still wondering is how to deal with overlapping numerical ranges (I think I would just throw an error) and how to represent intervals that go from minus infinity to some number or from some number to infinity. We could either allow things like `<3.5` or use `[-inf,3.5]` or similar. We could also just start with `<3.5` and `>3.5` for now and think about extending this to multiple ranges later.

Reply:
And another thing: What if we introduced a match-all key for enums like

```yaml
priority:
  road_class: {"motorway": 1, "*": 0.7}
```

So we can express things like I want to **prefer** the motorway without using priority > 1 (and the normalization we do because of this)? The match-all key would also be useful for speed_factor for example: reduce speed for all road types etc.

Reply:
Yeah, the "match all key" would make it possible to avoid the normalization for "**prefer**" and we could remove yet another root element "maxSpeedFallback". I will try to do the required changes.

Reply:
> Btw if we want to apply the same logic also for boolean, it would have to be get_off_bike: {"true": 0.4, "false": 0.2}

Will have to do this change too as there is no possibility to "**prefer**" a key then.

Reply:
> Will have to do this change too as there is no possibility to "prefer" a key then.

Ah yes this is because the factor must stay below zero, right? Btw in `SpeedCalculator` I thought it might make more sense to separate the cases by the type of the key instead of the value (`if (key instance of EnumEncodedValue)` instead of `if (value type of number)`).

Reply:
> Ah yes this is because the factor must stay below one, right?

Yes

> Btw in SpeedCalculator I thought it might make more sense to separate the cases by the type of the key instead of the value (if (key instance of EnumEncodedValue) instead of if (value type of number))

This is automatically the case with the new range stuff and "match all" key stuff, yes :)